### PR TITLE
RouteTable.Add overload with generic param of IDotvvmPresenter type

### PR DESCRIPTION
**Pull request by Leeemons**

### CODE REVIEW

File: src/DotVVM.Framework/Routing/DotvvmRouteTable.cs [hunk 1]
```diff
@@ -104,6 +104,21 @@ public void Add(string routeName, string url, string virtualPath, object default
...
+        /// Adds the specified route name.
+        /// </summary>
+        /// <typeparam name="T">Type of presenter</typeparam>
+        /// <param name="routeName">Name of the route.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="virtualPath">The virtual path of the Dothtml file.</param>
+        /// <param name="defaultValues">The default values.</param>
+        public void Add<T>(string routeName, string url, string virtualPath, object defaultValues = null) where T : IDotvvmPresenter
+        {
+            Func<IDotvvmPresenter> presenterFactory = presenterFactory = () => configuration.ServiceProvider.GetRequiredService<T>();
```
Comment (exyi) [inline src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: As @djanosik noticed, this does not work with request-scoped services.

File: src/DotVVM.Framework/Routing/DotvvmRouteTable.cs [hunk 2]
```diff
@@ -104,6 +104,21 @@ public void Add(string routeName, string url, string virtualPath, object default
...
 
+        /// <summary>
+        /// Adds the specified route name.
+        /// </summary>
+        /// <typeparam name="T">Type of presenter</typeparam>
+        /// <param name="routeName">Name of the route.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="virtualPath">The virtual path of the Dothtml file.</param>
+        /// <param name="defaultValues">The default values.</param>
+        public void Add<T>(string routeName, string url, string virtualPath, object defaultValues = null) where T : IDotvvmPresenter
```
Comment (exyi) [inline src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: The `virtualPath` parameter is completely redundant when the presenter is specified.

File: src/DotVVM.Framework/Routing/DotvvmRouteTable.cs [hunk 3]
```diff
@@ -104,6 +104,21 @@ public void Add(string routeName, string url, string virtualPath, object default
...
         }
 
+        /// <summary>
+        /// Adds the specified route name.
+        /// </summary>
+        /// <typeparam name="T">Type of presenter</typeparam>
+        /// <param name="routeName">Name of the route.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="virtualPath">The virtual path of the Dothtml file.</param>
+        /// <param name="defaultValues">The default values.</param>
```
Comment (exyi) [inline src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: When I look at the parameter name and the comment, I don't see any single bit of information that the comment gives me...

File: src/DotVVM.Framework/Routing/DotvvmRouteTable.cs (hunk 2 above)
Comment (exyi) [reply src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: As I told you in #463, we wanted a method with `Func<IServiceProvider, IDotvvmFactory>` parameters, this overload should be just a sugar for that (-> I'd make that an extension method). And I actually think that we don't need this overload, the presenterFactory can be quite easily used for this, see the discussion in #463

Comment (Leeemons) [reply src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: Why dont you like generic type?

**Comment by exyi** [comment]
> Why dont you like generic type?
It's not about generic types, the point that we need the overload with
the factory function, so this one becomes redundant.

**Comment by quigamdev** [comment]
I changed signature for DotvvmRouteTable.Add. There is breaking change, that requires change of docs.
@Leeemons  Please add tests for scoped lifecycle.

File: src/DotVVM.Framework/Routing/RouteBase.cs [hunk 4]
```diff
@@ -184,7 +184,7 @@ public static void AddOrUpdateParameterCollection(IDictionary<string, object> ta
         /// <summary>
         /// Processes the request.
         /// </summary>
-        public abstract IDotvvmPresenter GetPresenter();
+        public abstract IDotvvmPresenter GetPresenter(IServiceProvider provider);
```
Comment (exyi) [inline src/DotVVM.Framework/Routing/RouteBase.cs]: Btw: The comment is completely misleading.

Comment (quigamdev) [reply src/DotVVM.Framework/Routing/RouteBase.cs]: @Leeemons  Please change the comment.

File: src/DotVVM.Framework/Routing/DotvvmRouteTable.cs [hunk 5]
```diff
@@ -93,15 +93,37 @@ public void AddGroup(string groupName, string urlPrefix, string virtualPathPrefi
...
+        /// <summary>
+        /// Adds the specified route name.
+        /// </summary>
+        /// <param name="routeName">Name of the route.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="presenterType">The presenter factory.</param>
+        /// <param name="defaultValues">The default values.</param>
+        public void Add(string routeName, string url, Type presenterType, object defaultValues = null)
+        {
+            Func<IServiceProvider, IDotvvmPresenter> presenterFactory = provider => (IDotvvmPresenter)provider.GetService(presenterType);
```
Comment (tomasherceg) [inline src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: I would add the following check and provide the user with a meaningful error message:

`if (!typeof(IDotvvmPresenter).IsAssignableFrom(presenterType)) throw ...`

Comment (quigamdev) [reply src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: @tomasherceg  Please check the message text.

Comment (tomasherceg) [reply src/DotVVM.Framework/Routing/DotvvmRouteTable.cs]: I cannot see any message in the code, however I would use something like this:

"The presenter type {presenterType.ToString()} used in the route '{routeName}' does not implement IDotvvmPresenter!"

**Pull request merged**